batch.js

```diff
@@ -0,0 +1,317 @@
+const debugFactory = require('debug')
+
+const debug = debugFactory('graphile-upsert-batch')
+
+function PgMutationUpsertBatchPlugin(builder) {
+  builder.hook('GraphQLObjectType:fields', (fields, build, context) => {
+    const {
+      extend,
+      newWithHooks,
+      parseResolveInfo,
+      pgIntrospectionResultsByKind,
+      pgGetGqlTypeByTypeIdAndModifier,
+      pgGetGqlInputTypeByTypeIdAndModifier,
+      pgSql: sql,
+      gql2pg,
+      graphql: {
+        GraphQLObjectType,
+        GraphQLList,
+        GraphQLInputObjectType,
+        GraphQLNonNull,
+        GraphQLString,
+      },
+      pgColumnFilter,
+      inflection,
+      pgQueryFromResolveData: queryFromResolveData,
+      pgOmit: omit,
+      pgViaTemporaryTable: viaTemporaryTable,
+      describePgEntity,
+      sqlCommentByAddingTags,
+      pgField,
+    } = build
+    const {
+      scope: { isRootMutation },
+      fieldWithHooks,
+    } = context
+    if (!isRootMutation) {
+      return fields
+    }
+
+    return extend(
+      fields,
+      pgIntrospectionResultsByKind.class
+        .filter(table => !!table.namespace)
+        .filter(table => table.isSelectable)
+        .filter(table => table.isInsertable && !omit(table, 'create')) // we ignore update omits for simplicity
+        .reduce((memo, table) => {
+          const Table = pgGetGqlTypeByTypeIdAndModifier(table.type.id, null)
+          if (!Table) {
+            debug(
+              `There was no table type for table '${table.namespace.name}.${
+                table.name
+              }', so we're not generating an upsert batch mutation for it.`,
+            )
+            return memo
+          }
+          const TableInput = pgGetGqlInputTypeByTypeIdAndModifier(
+            table.type.id,
+            null,
+          )
+          if (!TableInput) {
+            debug(
+              `There was no input type for table '${table.namespace.name}.${
+                table.name
+              }', so we're going to omit it from the upsert batch mutation.`,
+            )
```

Comment:
Actually this looks deliberate? I don’t fully understand why though.

Reply:
ah, it's because this is based on [PgMutationCreatePlugin](https://github.com/graphile/graphile-engine/blob/a6b4dd9264017a70d20ca536212eac6ef7731651/packages/graphile-build-pg/src/plugins/PgMutationCreatePlugin.js#L71) which makes the same mistake!

Reply:
That’s because some tables require no values for insert (e.g. “create table foo (id int generated always as identity);”) but I’m not sure that makes sense for upsert?

Reply:
aaahh, well, I think there might be a theoretical case for that, but I've added the return as it's probably impractical.